### DISCUSSION TITLE
How to render a vertical bar chart

### DISCUSSION THREAD

**Discussion by codeminic** [oxyplot/oxyplot#1946] [Q&A] [answered]
I'm using OxyPlot to render a bar chart with SkiaSharp, for a PDF report. Currently, the bars are getting rendered horizontally, but i'd like them to be rendered vertically. 
I tried to swap the `AxisPosition` but this led to the following exception: 

```
OxyPlot exception: BarSeries requires a CategoryAxis on the Y Axis.
An exception of type System.Exception was thrown when updating the plot model.
at OxyPlot.Series.BarSeriesBase`1.GetCategoryAxis()
at 
OxyPlot.Series.BarSeriesBase`1.OxyPlot.Series.IBarSeries.get_CategoryAxis()
at OxyPlot.PlotModel.<>c.<UpdateBarSeriesManagers>b__238_1(IBarSeries s)
at System.Linq.Lookup`2.Create(IEnumerable`1 source, Func`2 keySelector, 
IEqualityCom
at System.Linq.GroupedEnumerable`2.GetEnumerator()
at OxyPlot.PlotModel.UpdateBarSeriesManagers()
at OxyPlot.PlotModel.OxyPlot.IPlotModel.Update(Boolean updateData)
```

I'm currently using this simplified example code snippet repo: 

```csharp
  var model = new PlotModel();

  var categoryAxis = new CategoryAxis { Position = AxisPosition.Left };
  var valueAxis1 = new LinearAxis
  {
      Title = "Value Axis 1",
      Position = AxisPosition.Bottom,
      MinimumPadding = 0.06,
      MaximumPadding = 0.06,
      ExtraGridlines = new[] {0d},
      Key = "x1"
  };
  model.Axes.Add(categoryAxis);
  model.Axes.Add(valueAxis1);

  var series = new List<BarSeries>
  {
      new BarSeries(),
      new BarSeries(),
  };

  var rnd = new Random(1);
  foreach (var s in series)
  {
      for (var i = 0; i < 4; i++)
          s.Items.Add(new BarItem() { Value = rnd.Next(0, 100) });
      model.Series.Add(s);
  }
```

This is the rendered bar chart:

![image](https://user-images.githubusercontent.com/16257869/194094451-9756c08a-b524-4c6e-ac96-4e42a6ad145b.png)

It looks like vertical bars were possible in the past using `ColumnSeries`, but this type of series has been removed. Does anyone know how this can be achieved? 

EDIT: I originally posted this on [SO](https://stackoverflow.com/questions/73958823/how-to-render-a-vertical-bar-chart-with-oxyplot)

**Comment by VisualMelon** [comment, marked as answer]
Set the keys explicitly on the axes and the series: the series' `YAxisKey` should correspond to the `Key` of the `CategoryAxis`; the series' `XAxisKey` should correspond to the `Key` of the value axis.

The approach of specifying the axes explicitly allows you to 'transpose' most series and annotations.

```
var model = new PlotModel();

var categoryAxis = new CategoryAxis { Position = AxisPosition.Bottom, Key = "y1" };
var valueAxis1 = new LinearAxis
{
    Title = "Value Axis 1",
    Position = AxisPosition.Left,
    MinimumPadding = 0.06,
    MaximumPadding = 0.06,
    ExtraGridlines = new[] { 0d },
    Key = "x1"
};
model.Axes.Add(categoryAxis);
model.Axes.Add(valueAxis1);

var series = new List<BarSeries>
{
  new BarSeries() { XAxisKey = "x1", YAxisKey = "y1" },
  new BarSeries() { XAxisKey = "x1", YAxisKey = "y1" },
};

var rnd = new Random(1);
foreach (var s in series)
{
    for (var i = 0; i < 4; i++)
        s.Items.Add(new BarItem() { Value = rnd.Next(0, 100) });
    model.Series.Add(s);
}
```

![image](https://user-images.githubusercontent.com/5646475/194105575-fa01efbd-14a1-4a71-9eaf-aa6037fd7d46.png)

**Reply by codeminic** [reply]
Thanks a lot!

**Comment by RoyalFireSpirit** [comment]
Thanks for the code @VisualMelon.  You say YAxisKey should correspond to the Key of the CategoryAxis. That seems confusing when the category axis is actually the X axis, so you're setting a property called YAxisKey to refer to the X axis (and vice versa).

Trying to understand what's going on here - is BarSeries.YAxisKey actually intended to always identify the category axis, no matter which axis that is?  If so, that would suggest the name of the property, YAxisKey, is a bit misleading. And ditto for BarSeries.XAxisKey.
